librz/il/il_validate.c

```diff
@@ -0,0 +1,798 @@
+		for (LocalPureVar *loc = local_pure_var_stack; loc; loc = loc->next) {
+			if (!strcmp(loc->name, args->v)) {
+				*sort_out = loc->sort;
+				return true;
+			}
+		}
+		VALIDATOR_ASSERT(false, "Local pure variable \"%s\" unbound at var op.\n", args->v);
+		return false;
+	}
+	default:
+		VALIDATOR_ASSERT(false, "Var op has invalid kind.\n");
+	}
+	return true;
+}
+
+VALIDATOR_PURE(bool_const) {
+	*sort_out = rz_il_sort_pure_bool();
+	return true;
+}
+
+VALIDATOR_PURE(bitv) {
+	RzBitVector *bv = op->op.bitv.value;
+	if (!bv) {
+		rz_strbuf_appendf(report_builder, "Bitvector in constant bitvector op is NULL.\n");
+		return false;
+	}
+	*sort_out = rz_il_sort_pure_bv(rz_bv_len(bv));
+	return true;
+}
+
+VALIDATOR_PURE(bitv_binop) {
+	RzILOpPure *x = op->op.add.x; // just add is fine, all ops in here use the same struct
+	RzILOpPure *y = op->op.add.y;
+	RzILSortPure sx;
+	VALIDATOR_DESCEND(x, &sx);
+	VALIDATOR_ASSERT(sx.type == RZ_IL_TYPE_PURE_BITVECTOR, "Left operand of %s op is not a bitvector.\n", rz_il_op_pure_code_stringify(op->code));
+	RzILSortPure sy;
+	VALIDATOR_DESCEND(y, &sy);
+	VALIDATOR_ASSERT(sy.type == RZ_IL_TYPE_PURE_BITVECTOR, "Right operand of %s op is not a bitvector.\n", rz_il_op_pure_code_stringify(op->code));
+	VALIDATOR_ASSERT(sx.props.bv.length == sy.props.bv.length, "Operand sizes of %s op do not agree: %u vs. %u.\n",
+		rz_il_op_pure_code_stringify(op->code), (unsigned int)sx.props.bv.length, (unsigned int)sy.props.bv.length);
+	*sort_out = sx;
+	return true;
+}
+
+VALIDATOR_PURE(ite) {
+	RzILOpArgsIte *args = &op->op.ite;
+	RzILSortPure sc;
+	VALIDATOR_DESCEND(args->condition, &sc);
+	VALIDATOR_ASSERT(sc.type == RZ_IL_TYPE_PURE_BOOL, "Condition of ite op is not boolean.\n");
+	RzILSortPure sx;
+	VALIDATOR_DESCEND(args->x, &sx);
+	RzILSortPure sy;
+	VALIDATOR_DESCEND(args->y, &sy);
+	if (!rz_il_sort_pure_eq(sx, sy)) {
+		char *sxs = rz_il_sort_pure_stringify(sx);
+		char *sys = rz_il_sort_pure_stringify(sy);
+		rz_strbuf_appendf(report_builder, "Types of ite branches do not agree: %s vs. %s.\n",
+			rz_str_get_null(sxs), rz_str_get_null(sys));
+		free(sxs);
+		free(sys);
+		return false;
+	}
+	*sort_out = sx;
+	return true;
+}
+
+VALIDATOR_PURE(let) {
+	RzILOpArgsLet *args = &op->op.let;
+	VALIDATOR_ASSERT(args->name, "Var name of let op is NULL.\n");
+	VALIDATOR_ASSERT(args->exp, "Expression of let op is NULL.\n");
+	VALIDATOR_ASSERT(args->body, "Body of let op is NULL.\n");
+	RzILSortPure sort;
+	VALIDATOR_DESCEND(args->exp, &sort);
+	LocalPureVar var = {
+		.name = args->name,
+		.sort = sort,
+		.next = local_pure_var_stack
+	};
+	return validate_pure(args->body, sort_out, report_builder, ctx, &var);
+}
+
+VALIDATOR_PURE(inv) {
+	RzILOpArgsBoolInv *args = &op->op.boolinv;
+	RzILSortPure sort;
+	VALIDATOR_DESCEND(args->x, &sort);
+	VALIDATOR_ASSERT(sort.type == RZ_IL_TYPE_PURE_BOOL, "Operand of boolean inv op is not boolean.\n");
+	*sort_out = rz_il_sort_pure_bool();
+	return true;
+}
+
+VALIDATOR_PURE(bool_binop) {
```

Comment:
For me, it's obvious that we are dealing with binary ops, then unary ops, etc. But for new devs it's probably worth to add a brief description for those validation functions.